flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -742,10 +743,38 @@ private void notifyFatalError(String message, Throwable cause) {
 	}
 
 	// ----------------------------------------------------------------------------------------------------------------
-	//  Canceling / Failing the task from the outside
+	//  Stopping / Canceling / Failing the task from the outside
 	// ----------------------------------------------------------------------------------------------------------------
 
 	/**
+	 * Stops the executing task by calling {@link Stoppable#stop()}.
+	 * <p>
+	 * This method never blocks.
+	 * </p>
+	 * 
+	 * @throws UnsupportedOperationException
+	 *             if the {@link AbstractInvokable} does not implement {@link Stoppable}
+	 */
+	public void stopExecution() throws UnsupportedOperationException {
+		LOG.info("Attempting to stop task " + taskNameWithSubtask);
+		if(this.invokable instanceof Stoppable) {
+			Runnable runnable = new Runnable() {
+				@Override
+				public void run() {
+					try {
+						((Stoppable)Task.this.invokable).stop();
+					} catch(RuntimeException e) {
+						LOG.error("Stopping task " + taskNameWithSubtask + " failed.", e);
```

Comment:
Shouldn't a `RuntimeException` cause the task to fail? We should notify the `JobManager` about this and fail the corresponding task. E.g. we could send a `FailTask` message to the `TaskManager` to fail the task.